src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.Unix.cs

```diff
@@ -41,26 +42,49 @@ internal static string Normalize(string strInput, NormalizationForm normalizatio
 
             ValidateArguments(strInput, normalizationForm);
 
-            char[] buf = new char[strInput.Length];
-
-            for (int attempts = 2; attempts > 0; attempts--)
+            char[]? toReturn = null;
+            try
             {
-                int realLen = Interop.Globalization.NormalizeString(normalizationForm, strInput, strInput.Length, buf, buf.Length);
+                int realLen = strInput.Length;
 
-                if (realLen == -1)
+                for (int attempt = 0; attempt < 2; attempt++)
                 {
-                    throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    if (toReturn != null)
+                    {
+                        // Clear toReturn first to ensure we don't return the same buffer twice
+                        char[] temp = toReturn;
+                        toReturn = null;
+                        ArrayPool<char>.Shared.Return(temp);
+                    }
+
+                    Span<char> buffer = realLen <= 512
+                        ? stackalloc char[512]
```

Comment:
>stackalloc char[512] [](start = 26, length = 20)

should we initialize the stackalloc outside the loop once? we don't

Reply:
Changed